### PR TITLE
AuthType Session Fix

### LINKED ISSUE #68: openid (keycloak) auth: kibana does not refresh access token
Version: ES / Kibana 1.0.1

Reproduce:
- Setup keycloak as openid provider in both securityconfig and kibana configuration
- Successfully authentication
- Wait 1 minute on the main page (or until session expiration)
- Try to browse anywhere
- Kibana will fail to respond unless page is refreshed

What it seems to be happening is Kibana responds with a 301 as you might expect. It then does a CORS OPTIONS request (which is enabled and allowed in my keycloak). It then does a GET request against the keycloak auth endpoint which returns a 200 but does not renew the existing session but instead assumes it is logged out. What I believe is happening is cookies are not being sent in this final request. After keycloak returns a 200 Kibana will fail to respond unless you browse to timelion or the security panel.

### PR DESCRIPTION
**Pull request by shamsalmon**
This pull request resolves issues with session timeouts and ajax requests for a openid connect endpoint resulting in CORS issues.

Tracking the fix to https://github.com/opendistro-for-elasticsearch/security-kibana-plugin/issues/68

Amazon has fixed most of the changes related to session timeouts and redirects. The only thing that does not work is XHR requests currently once the session has expired. To replicate the issue:

1. Go to search
2. Wait for session to expire
3. Press refresh

It will then throw an exception related to authentication (as of ODFE 1.3.0). This is due to an angular bug which is addressed in this PR.

### CODE REVIEW

**Comment by shamsalmon** [comment]
I tracked down what is happening here... if I delete the cookies so cookie validator is never ran then we get to the point where headers are considered (from an ajax request) and a 401 is sent. 

You will only notice this happening if your session timeouts are very low. Mine is set to 1 minute.

Every instance of a recent version of Keyclock I have stood up will not respond to CORS the way this expects even with web origins added, and this will be broken for you. 

This is fixable by moving the logic where its checking the header above where the cookie validation is sending it to onUnathenticated.

**Comment by shamsalmon** [comment]
Remediate session issues by using the code above and setting `opendistro_security.cookie.ttl` to the same expration as the setting in your keycloak client.

What this will do is skip the cookievalidation errors I posted about above and instead validate your ajax request and send a 401 if it is one.

It will then handle the 401 on the frontend and redirect you through a login flow that does not require CORS.

What this does not fix:
POST requests after the session expires. If you wait for the session and cookie to expire you will see POST requests returning a 401 but not redirecting. This is because the UI error handler that handles the login flow for 401s does not account for POST requests. What it should do is reauth you but instead it just throws an auth error. That needs to be fixed.

**Comment by shamsalmon** [comment]
@hardik-k-shah

**Comment by shamsalmon** [comment]
This is now fixed for post requests with the latest commit. The flow looks like this if you are going through XHR:

Request sent -> Backend sends 401 if session is expired and cookie has not been invalidated (cookie expiration (ttl setting) must be higher than session expiration) -> Frontend captures 401 and does a window.location redirect forcing a reauthentication window.

This *should* also fix SAML and every other provider that relies on the same workflow.

**Comment by rofreytag** [comment]
+1 Please include this in an upcoming patch. 

We love using kibana behind SAML, the session loss errors are really annoying.

**Comment by shamsalmon** [comment]
My latest change fixes the json parsing issue on the frontend handler. I removed formatting stuff and things odfe 1.2 security plugin fixes.

**Comment by agarwalgovind** [comment]
Hi Sam,

Are the latest changes involves using refresh_token for access_token renewal when we use odfe with OpenID and Keycloak or currently there is some other way to renew the access_token till SSO session is valid.

**Comment by shamsalmon** [comment]
> Hi Sam,
> 
> Are the latest changes involves using refresh_token for access_token renewal when we use odfe with OpenID and Keycloak or currently there is some other way to renew the access_token till SSO session is valid.

This change only forces a redirect if the session is expired. Access / Refresh token changes will require some significant changes.

**Comment by chasebrewsky** [comment]
+1 We currently have an OpenID authentication token that expires every 5 minutes for security purposes. Without this change we have to refresh the page constantly to keep the session alive.

**Comment by shamsalmon** [comment]
> +1 We currently have an OpenID authentication token that expires every 5 minutes for security purposes. Without this change we have to refresh the page constantly to keep the session valid. We also have the same issue that agarwalgovind has, where we are issuing a refresh token but it never gets used to refresh the access token.

I am no longer currently using Open Distro for my work. I looked into the change required to make a request for a new token and it doesn't seem overly complicated. Here is a link to the auth api that Open Distro uses. https://github.com/hapijs/bell/blob/master/API.md. There is a section there about refresh tokens which you can use to generate a new token. I was planning on implementing this at some point but I haven't had the time unfortunately.

**Comment by chasebrewsky** [comment]
> I am no longer currently using Open Distro for my work. I looked into the change required to make a request for a new token and it doesn't seem overly complicated. Here is a link to the auth api that Open Distro uses. https://github.com/hapijs/bell/blob/master/API.md. There is a section there about refresh tokens which you can use to generate a new token. I was planning on implementing this at some point but I haven't had the time unfortunately.

That's alright, after reviewing a lot of the pull requests that have been in the queue vs what is actually getting pulled in it looks like they've been prioritizing pull requests from internal employees over external requests for some time now.

I also looked at the code last night and noticed something similar. The biggest thing I noticed was that each full page load redirected to the OpenID authentication endpoint instead of tracking the expiration of the actual token given in the token response, attempting a refresh, and continuing on. I'm guessing this is because the functionality didn't exactly fit into the interface for the plugin or requires implementation specific middleware?

Did you end up using just plain old Elasticsearch in the end?

**Comment by shamsalmon** [comment]
> > I am no longer currently using Open Distro for my work. I looked into the change required to make a request for a new token and it doesn't seem overly complicated. Here is a link to the auth api that Open Distro uses. https://github.com/hapijs/bell/blob/master/API.md. There is a section there about refresh tokens which you can use to generate a new token. I was planning on implementing this at some point but I haven't had the time unfortunately.
> 
> That's alright, after reviewing a lot of the pull requests that have been in the queue vs what is actually getting pulled in it looks like they've been prioritizing pull requests from internal employees over external requests for some time now.
> 
> I also looked at the code last night and noticed something similar. The biggest thing I noticed was that each full page load redirected to the OpenID authentication endpoint instead of tracking the expiration of the actual token given in the token response, attempting a refresh, and continuing on. I'm guessing this is because the functionality didn't exactly fit into the interface for the plugin or requires implementation specific middleware?
> 
> Did you end up using just plain old Elasticsearch in the end?

No although we have plans to implement Open Distro within the next few months. Its just not current at the top of my to-dos. 

The behavior you are seeing is because refresh tokens are not utilized at all as far as I can tell. The bell oauth library does not seem to automate that as much as I am sure they would hope.

**Comment by james-od** [comment]
Another +1 to this
Our engineers are finding it frustrating to have to re-navigate to their indexes several times a day due to session timeouts, getting this fix in would be fantastic.
Thanks!

**Comment by fatalglitch** [comment]
+1 Let's get this moving to fix this issue

**Comment by agarwalgovind** [comment]
Hello,

I have a fix for the issue. Will be raising a PR. The fix is for ODFE - 0.9 w.r.t to Kibana 6.7.
After approval and thorough review the fix can be patched to higher version also.

File: public/chrome/configuration/enable_configuration.js
Comment (seraphjiang) [inline public/chrome/configuration/enable_configuration.js:83]: use ```let``` than ```var```

**Review by seraphjiang — CHANGES_REQUESTED** [review]
would you rebase from master

**Comment by seraphjiang** [comment]
Close PR as no response, feel free to reopen

**Comment by rofreytag** [comment]
Oh no, so sad about this getting closed.

I hope either the original author or another maintainer can re-open this. We have been waiting for a fix on this a while now.  
I can only repeat what @james-od said, it is a time-burner to constantly need to refresh or re-login to kibana due to this issue.

**Comment by agarwalgovind** [comment]
Hello,

There is a PR for the above issue - [](https://github.com/opendistro-for-elasticsearch/security-kibana-plugin/pull/217.) It being tracked under a new ticket.

**Pull request closed without merging**
